src/Microsoft.AspNetCore.ResponseCompression/ResponseCompressionProvider.cs

```diff
@@ -62,42 +72,74 @@ public ResponseCompressionProvider(IServiceProvider services, IOptions<ResponseC
-                var sorted = unsorted
-                    .Where(s => s.Quality.GetValueOrDefault(1) > 0)
-                    .OrderByDescending(s => s.Quality.GetValueOrDefault(1));
+                return null;
+            }
 
-                foreach (var encoding in sorted)
+            if (StringWithQualityHeaderValue.TryParseList(accept, out var encodings))
+            {
+                if (encodings.Count == 0)
                 {
-                    // There will rarely be more than three providers, and there's only one by default
-                    foreach (var provider in _providers)
+                    return null;
+                }
+
+                var candidates = new HashSet<ProviderCandidate>();
+
+                foreach (var encoding in encodings)
+                {
+                    var encodingName = encoding.Value;
+                    var quality = encoding.Quality.GetValueOrDefault(1);
+
+                    if (quality < double.Epsilon)
                     {
-                        if (StringSegment.Equals(provider.EncodingName, encoding.Value, StringComparison.OrdinalIgnoreCase))
+                        continue;
+                    }
+
+                    for (int i = 0; i < _providers.Length; i++)
+                    {
+                        var provider = _providers[i];
+
+                        if (StringSegment.Equals(provider.EncodingName, encodingName, StringComparison.OrdinalIgnoreCase))
                         {
-                            return provider;
+                            candidates.Add(new ProviderCandidate(provider.EncodingName, quality, i, provider));
                         }
                     }
 
                     // Uncommon but valid options
-                    if (StringSegment.Equals("*", encoding.Value, StringComparison.Ordinal))
+                    if (StringSegment.Equals("*", encodingName, StringComparison.Ordinal))
                     {
-                        // Any
-                        return _providers[0];
+                        for (int i = 0; i < _providers.Length; i++)
+                        {
+                            var provider = _providers[i];
+                            
+                            // Any provider is a candidate.
+                            candidates.Add(new ProviderCandidate(provider.EncodingName, quality, i, provider));
+                        }
+
+                        break;
                     }
-                    if (StringSegment.Equals("identity", encoding.Value, StringComparison.OrdinalIgnoreCase))
+
+                    if (StringSegment.Equals("identity", encodingName, StringComparison.OrdinalIgnoreCase))
                     {
-                        // No compression
-                        return null;
+                        candidates.Add(new ProviderCandidate(encodingName.Value, quality, int.MaxValue, null));
```

Comment:
We should add a comment here about what each of these values mean.

Reply:
The arguments?

Reply:
Yeah for the arguments. Or add named params.